core/src/main/java/org/apache/calcite/sql/fun/SqlDialectOperatorTableFactory.java

```diff
@@ -0,0 +1,247 @@
+   * <i>search_chars</i> replaced by its corresponding character in <i>replacement_chars</i>.
+   *
+   * <p>It is not defined in the SQL standard, but occurs in Oracle and PostgreSQL.
+   */
+  @SqlDialect(dialects = {Dialect.ORACLE, Dialect.POSTGRES})
+  public static final SqlFunction TRANSLATE3 = new SqlTranslate3Function();
+
+  /**
+   * Returns the sql dialect operator table factory, creating it if necessary.
+   */
+  public static synchronized SqlDialectOperatorTableFactory instance() {
+    if (instance == null) {
+      // Creates and initializes this table factory.
+      instance = new SqlDialectOperatorTableFactory();
+      instance.init();
+    }
+    return instance;
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  /**
+   * Performs post-constructor initialization of an operator table. It can't
+   * be part of the constructor, because the subclass constructor needs to
+   * complete first.
+   */
+  public final void init() {
+    // Initialize the dialect to SqlOperator mapping.
+    for (Dialect dialect : Dialect.values()) {
+      dialectSqlOperatorTableMapping.put(dialect, new ListSqlOperatorTable());
+    }
+
+    // Use reflection to register the expressions stored in public fields.
+    for (Field field : getClass().getFields()) {
+      try {
+        if (SqlFunction.class.isAssignableFrom(field.getType())) {
+          SqlFunction op = (SqlFunction) field.get(this);
+          if (op != null) {
+            register(op, getSqlDialectsFromField(op.getName(), field));
+          }
+        } else if (
+            SqlOperator.class.isAssignableFrom(field.getType())) {
+          SqlOperator op = (SqlOperator) field.get(this);
+          register(op, getSqlDialectsFromField(op.getName(), field));
+        }
+      } catch (IllegalArgumentException | IllegalAccessException e) {
+        Util.throwIfUnchecked(e.getCause());
+        throw new RuntimeException(e.getCause());
+      }
+    }
+    this.initialized = true;
+  }
+
+  /** Extract supported sql dialects from the field annotation. */
+  private Dialect[] getSqlDialectsFromField(String operatorName, Field field) {
+    SqlDialect sqlDialect = field.getAnnotation(SqlDialect.class);
+    assert sqlDialect != null : "Must annotate with SqlDialect annotation for operator "
+        + operatorName;
+    Dialect[] dialects = sqlDialect.dialects();
+    assert dialects.length > 0 : "Must specify at least one dialect for operator "
+        + operatorName;
+    return dialects;
+  }
+
+  private void register(SqlOperator operator, Dialect[] dialects) {
+    for (Dialect dialect : dialects) {
+      dialectSqlOperatorTableMapping.get(dialect).add(operator);
+    }
+  }
+
+  /** Returns an immutable sql operator table by specific sql dialect. */
+  public ImmutableSqlOperatorTable getOperatorTable(Dialect dialect) {
+    assert initialized : "Please invoke init() to initialize this operator table first.";
```

Comment:
The type of return value can be `SqlOperatorTable` instead of `ImmutableSqlOperatorTable`.

Reply:
I agree, will change to SqlOperatorTable.